### PR TITLE
add convenience application factory for self-hosting

### PR DESCRIPTION
**Pull request by ZzzzHui**

### CODE REVIEW

File: contracts/dapp/ISelfHostedApplicationFactory.sol [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+import {Authority} from "../consensus/authority/Authority.sol";
+import {IAuthorityFactory} from "../consensus/authority/IAuthorityFactory.sol";
+import {Application} from "./Application.sol";
+import {IApplicationFactory} from "./IApplicationFactory.sol";
+import {IInputBox} from "../inputs/IInputBox.sol";
+import {IPortal} from "../portals/IPortal.sol";
+
+/// @title Self-hosted Application Factory interface
+interface ISelfHostedApplicationFactory {
+    /// @notice Get the factory used to deploy `Authority` contract
```
Comment (guidanoli) [inline contracts/dapp/ISelfHostedApplicationFactory.sol:15]:
```suggestion
    /// @notice Get the factory used to deploy `Authority` contracts
```

File: contracts/dapp/SelfHostedApplicationFactory.sol [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        Application application = _applicationFactory.newApplication(
+            authority,
+            inputBox,
+            portals,
+            appOwner,
+            templateHash,
+            salt
+        );
+
+        return (application, authority);
```
Comment (guidanoli) [inline contracts/dapp/SelfHostedApplicationFactory.sol:72]: You could make the return values named, just like you did in `calculateAddresses`, and just assign the newly-created contracts to them. With this, you won't need to declare them, won't need the `return` statement, and the first statement might even fit in one line?

**Review by guidanoli — COMMENTED** [review]
Besides a small change, looks good!

File: contracts/dapp/SelfHostedApplicationFactory.sol (hunk 2 above)
Comment (ZzzzHui) [reply contracts/dapp/SelfHostedApplicationFactory.sol:72]: Hey Gui. I modified the `calculateAddresses` so there's no suffix `_` for return variables. I think suffix `_` for return variables make sense if there's prefix `_` for function parameters.  But we are using prefix `_` as internal state variables.

Comment (guidanoli) [reply contracts/dapp/SelfHostedApplicationFactory.sol:72]: Oh, I didn't notice the `_` suffix. Yeah, we're no longer doing that. But I think we can still name return values, like `application` and `authority`.

Comment (ZzzzHui) [reply contracts/dapp/SelfHostedApplicationFactory.sol]: True. Done. 👍

**Review by guidanoli — APPROVED** [review]
Great job, LGTM!

**Pull request merged**